From 69a2dbca759a46c953a3151320343925a2db7046 Mon Sep 17 00:00:00 2001
From: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
Date: Mon, 7 Feb 2022 14:52:55 -0800
Subject: [PATCH] ath12k: WAR for Ethernet mode

Use different bank id for handling frames like eapol/arp which
are in native wifi format instead of ethernet as configured in
primary bank.

Signed-off-by: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h  |  1 +
 drivers/net/wireless/ath/ath12k/dp.c    | 32 ++++++++++++++++---------
 drivers/net/wireless/ath/ath12k/dp_tx.c | 20 +++++++++++++---
 drivers/net/wireless/ath/ath12k/mac.c   |  2 ++
 4 files changed, 41 insertions(+), 14 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 56c007c08d15..f4e7a8918bea 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -232,6 +232,7 @@ struct ath12k_vif {
 	struct ieee80211_vif *vif;
 
 	int bank_id;
+	int bank_id_custom;
 	u8 vdev_id_check_en;
 
 	u16 tx_seq_no;
diff --git a/drivers/net/wireless/ath/ath12k/dp.c b/drivers/net/wireless/ath/ath12k/dp.c
index d1abf8c9b4e0..6da686a6308b 100644
--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -316,18 +316,25 @@ int ath12k_dp_srng_setup(struct ath12k_base *ab, struct dp_srng *ring,
 
 static
 void ath12k_dp_tx_get_vdev_bank_config(struct ath12k_base *ab, struct ath12k_vif *arvif,
-				       u32 *bank_config)
+				       u32 *bank_config, bool custom)
 {
 	enum hal_encrypt_type encrypt_type = 0;
+	enum hal_tcl_encap_type encap_type = 0;
 
-	/* Only valid for raw frames with HW crypto enabled.
-	 * With SW crypto, mac80211 sets key per packet
-	 */
-	if (arvif->tx_encap_type == HAL_TCL_ENCAP_TYPE_RAW &&
-	    test_bit(ATH12K_FLAG_HW_CRYPTO_DISABLED, &ab->dev_flags))
-		encrypt_type = ath12k_dp_tx_get_encrypt_type(arvif->key_cipher);
+	if (custom) {
+		encap_type = HAL_TCL_ENCAP_TYPE_NATIVE_WIFI;
+		encrypt_type = HAL_ENCRYPT_TYPE_OPEN;
+	} else {
+		encap_type = arvif->tx_encap_type;
+		/* Only valid for raw frames with HW crypto enabled.
+		 * With SW crypto, mac80211 sets key per packet
+		 */
+		if (arvif->tx_encap_type == HAL_TCL_ENCAP_TYPE_RAW &&
+		    test_bit(ATH12K_FLAG_HW_CRYPTO_DISABLED, &ab->dev_flags))
+			encrypt_type = ath12k_dp_tx_get_encrypt_type(arvif->key_cipher);
+	}
 
-	*bank_config |= FIELD_PREP(HAL_TX_BANK_CONFIG_ENCAP_TYPE, arvif->tx_encap_type) |
+	*bank_config |= FIELD_PREP(HAL_TX_BANK_CONFIG_ENCAP_TYPE, encap_type) |
 			FIELD_PREP(HAL_TX_BANK_CONFIG_ENCRYPT_TYPE, encrypt_type);
 	*bank_config |= FIELD_PREP(HAL_TX_BANK_CONFIG_SRC_BUFFER_SWAP, 0) |
 			FIELD_PREP(HAL_TX_BANK_CONFIG_LINK_META_SWAP, 0) |
@@ -350,14 +357,14 @@ void ath12k_dp_tx_get_vdev_bank_config(struct ath12k_base *ab, struct ath12k_vif
 }
 
 static int ath12k_dp_tx_get_bank_profile(struct ath12k_base *ab, struct ath12k_vif *arvif,
-					 struct ath12k_dp *dp)
+					 struct ath12k_dp *dp, bool custom)
 {
 	int bank_id = DP_INVALID_BANK_ID;
 	int i;
 	u32 bank_config = 0;
 
 	/* convert vdev params into hal_tx_bank_config */
-	ath12k_dp_tx_get_vdev_bank_config(ab, arvif, &bank_config);
+	ath12k_dp_tx_get_vdev_bank_config(ab, arvif, &bank_config, custom);
 
 	spin_lock_bh(&dp->tx_bank_lock);
 	/* TODO: implement using idr kernel framework*/
@@ -1091,7 +1098,10 @@ void ath12k_dp_vdev_tx_attach(struct ath12k *ar, struct ath12k_vif *arvif)
 
 	ath12k_dp_update_vdev_search(arvif);
 	arvif->vdev_id_check_en = true;
-	arvif->bank_id = ath12k_dp_tx_get_bank_profile(ab, arvif, &ab->dp);
+	arvif->bank_id = ath12k_dp_tx_get_bank_profile(ab, arvif, &ab->dp, false);
+	if (arvif->tx_encap_type != HAL_TCL_ENCAP_TYPE_NATIVE_WIFI)
+		arvif->bank_id_custom = ath12k_dp_tx_get_bank_profile(ab, arvif,
+								      &ab->dp, true);
 
 	/* TODO: error path for bank id failure */
 	if (arvif->bank_id == DP_INVALID_BANK_ID) {
diff --git a/drivers/net/wireless/ath/ath12k/dp_tx.c b/drivers/net/wireless/ath/ath12k/dp_tx.c
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -172,7 +176,7 @@ tcl_ring_sel:
 
 	ti.bank_id = arvif->bank_id;
 	ti.meta_data_flags = arvif->tcl_metadata;
-
+#if 0
 	if (arvif->tx_encap_type == HAL_TCL_ENCAP_TYPE_RAW &&
 	    test_bit(ATH12K_FLAG_HW_CRYPTO_DISABLED, &ar->ab->dev_flags)) {
 		if (skb_cb->flags & ATH12K_SKB_CIPHER_SET) {
@@ -185,7 +189,18 @@ tcl_ring_sel:
 			ti.encrypt_type = HAL_ENCRYPT_TYPE_OPEN;
 		}
 
-		msdu_ext_desc = true;
+		msdu_ext_desc = false;
+	}
+#endif
+	ti.encap_type = ath12k_dp_tx_get_encap_type(arvif, skb);
+	if ((arvif->tx_encap_type == HAL_TCL_ENCAP_TYPE_ETHERNET &&
+	    !(info->flags & IEEE80211_TX_CTL_HW_80211_ENCAP)) ||
+	    skb->protocol == cpu_to_be16(ETH_P_PAE)) {
+		ti.bank_id = arvif->bank_id_custom;
+
+		if (skb_cb->flags & ATH12K_SKB_CIPHER_SET &&
+		    ieee80211_has_protected(hdr->frame_control))
+			skb_put(skb, IEEE80211_CCMP_MIC_LEN);
 	}
 
 	ti.encap_type = ath12k_dp_tx_get_encap_type(arvif, skb);
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 74c169bb69d2..1f49b395ccb9 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -5619,6 +5619,8 @@ err_vdev_del:
 
 	ath12k_mac_vif_unref(&ab->dp, vif);
 	ath12k_dp_tx_put_bank_profile(&ab->dp, arvif->bank_id);
+	if (arvif->tx_encap_type != HAL_TCL_ENCAP_TYPE_NATIVE_WIFI)
+		ath12k_dp_tx_put_bank_profile(&ab->dp, arvif->bank_id_custom);
 
 	/* Recalc txpower for remaining vdev */
 	ath12k_mac_txpower_recalc(ar);
-- 
2.17.1

